File: go/pkg/common/engine/etcd/stream.go
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2025. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Package etcd implements engine.Engine
package etcd

import (
	"io"
	"yuanrong.org/kernel/pkg/common/engine"
)

type etcdStmt struct {
	fn      func() ([]interface{}, error)
	filters []engine.FilterFunc
}

// Filter implements engine.PrepareStmt
func (s *etcdStmt) Filter(filter engine.FilterFunc) engine.PrepareStmt {
	s.filters = append(s.filters, filter)
	return s
}

// Execute implements engine.PrepareStmt
func (s *etcdStmt) Execute() (engine.Stream, error) {
	vs, err := s.fn()
	if err != nil {
		return nil, err
	}

	var res []interface{}
outer:
	for _, v := range vs {
		for _, filter := range s.filters {
			if !filter(v) {
				continue outer
			}
		}
		res = append(res, v)
	}
	return &etcdStream{vs: res}, nil
}

type etcdStream struct {
	vs  []interface{}
	pos int
}

// Next implements engine.Stream
func (s *etcdStream) Next() (interface{}, error) {
	defer func() {
		s.pos++
	}()

	if s.pos == len(s.vs) {
		return nil, io.EOF
	}
	return s.vs[s.pos], nil
}

File: api/java/function-common/src/main/java/org/yuanrong/services/runtime/action/Stream.java
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2025-2026. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.yuanrong.services.runtime.action;

import org.yuanrong.errorcode.ErrorInfo;
import org.yuanrong.exception.LibRuntimeException;
import org.yuanrong.exception.YRException;
import org.yuanrong.jni.LibRuntime;
import org.yuanrong.runtime.util.Utils;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonObject;

import lombok.Setter;

import java.nio.charset.StandardCharsets;

/**
 * The stream handler of context, used to do some stream operation in FaaS mode.
 *
 * @since 2025-09-11
 */
public class Stream {
    private static final Gson GSON = new GsonBuilder().serializeNulls().create();

    @Setter
    private int maxWriteSize = 4 * 1024 * 1024;

    @Setter
    private String requestId = "";

    @Setter
    private String instanceId = "";

    /**
     * write jsonObject.
     *
     * @param jsonObject jsonObject
     * @throws YRException exception
     */
    public void write(JsonObject jsonObject) throws YRException {
        String jsonStr = GSON.toJson(jsonObject);
        int writeSizeInBytes = 0;
        if (jsonStr != null) {
            writeSizeInBytes = jsonStr.getBytes(StandardCharsets.UTF_8).length;
        }
        if (writeSizeInBytes > maxWriteSize) {
            throw new IllegalArgumentException("Write data size is larger than " + maxWriteSize);
        }
        ErrorInfo errorInfo;
        try {
            errorInfo = LibRuntime.streamWrite(jsonStr, requestId, instanceId);
        } catch (LibRuntimeException e) {
            throw new YRException(e.getErrorCode(), e.getModuleCode(), e.getMessage());
        }
        Utils.checkErrorAndThrow(errorInfo, "Write data failed");
    }
}
